Slow `compact_allocate`

I'm seeing what I think is a pathological performance in `compact_allocate`.

The following program allocates ~1GB of memory and does compaction:

```
let () = Gc.set { (Gc.get ()) with minor_heap_size = 1048576 }
let v = List.init (1_000_000_000 / (8 * 3)) (fun i -> i)
let () = Gc.compact ()
let (_ : _) = Sys.opaque_identity v
```

The compaction takes ~5s, almost all of the time taken by `compact_allocate`.
Unfortunately, this benchmark is fragile: in particular I can't reproduce the issue without the [Gc.set] call, even though it doesn't matter what I set the minor heap size to. (it only matters that it's different from the initial value).

Here's a perf report (~200 samples in total here):

```
       │    00000000005a3f30 <compact_allocate>:
       │    compact_allocate():
       │      mov    compact_fl,%rax
       │      mov    -0x30(%rax),%rcx
       │      mov    -0x28(%rax),%rdx
       │      sub    %rcx,%rdx
       │      cmp    $0xf,%rdx
       │    ↓ ja     3c
       │      nop
       │20:   mov    -0x20(%rax),%rax
       │      mov    -0x30(%rax),%rcx
       │      mov    -0x28(%rax),%rdx
       │      sub    %rcx,%rdx
       │      cmp    $0xf,%rdx
       │    ↑ jbe    20
       │      mov    %rax,compact_fl
       │3c:   cmp    %rdx,%rdi
       │    ↓ jbe    5c
       │      nop
       │48:┌─→mov    -0x20(%rax),%rax
       │   │  mov    -0x30(%rax),%rcx
 91.70 │   │  mov    -0x28(%rax),%rdx
       │   │  sub    %rcx,%rdx
       │   │  cmp    %rdi,%rdx
  8.30 │   └──jb     48
       │5c:   add    %rcx,%rdi
       │      mov    %rdi,-0x30(%rax)
       │      add    %rcx,%rax
       │    ← retq
```

We looked at this with Leo (@lpw25) and our conjecture is that the allocation algorithm suffers from the known first-fit performance problem: for every element of the list, the allocation algorithm scans all blocks on the free list. In our case we have ~45 blocks, so allocation is slower than what it should be by that factor.

One improvement idea, suggested by Leo, is to implement next-fit instead of first-fit. 
If we're worried about sacrificing even the smallest bit of compactness, we can implement a tweak to the next-fit algorithm: maintain a pointer in the free list which tells us which blocks have been used already, and which ones are completely free (therefore subject to being dropped). Then we make sure to avoid touching a completely free block unless we've scanned through all blocks that have been used.

I can also think of a more efficient (in the worst-case) implementation of first-fit semantics, where instead of scanning the free list from the beginning every time, we scan it backwards from the point where the previous scan ended (we need to keep track of extra info for this: the monotonically increasing subsequence of the prefix). However, next-fit seems just simpler.

<!--
Welcome to OCaml's Issue tracker!

OCaml's developers use this tracker for bugs and feature requests only, rather
than user support.

If you have questions about *using* OCaml, please ask at
https://discuss.ocaml.org (more people read Discuss than this tracker, and
you'll get confirmation of whether you've really found a bug or need a new
feature).

If your error came from the OCaml package manager, opam, (messages beginning
`[ERROR] The compilation of ...`), please start at
https://github.com/ocaml/opam-repository/issues/new.

Some libraries and tools which used to be part of OCaml are now maintained
separately. Please post questions about Graphics, Num, camlp4, LablTk, CamlDBM
or OCamlTopWin on Discuss or on their respective issue trackers (see [README.adoc](https://github.com/ocaml/ocaml/blob/trunk/README.adoc#separately-maintained-components)
for a full list).
-->

--- a/Changes
+++ b/Changes
@@ -10,7 +10,7 @@ OCaml 4.14 maintenance version
 
 - #12264, #12289: Fix compact_allocate to avoid a pathological case
   that causes very slow compaction.
-  (Damien Doligez, report by Arseniy Alekseyev, review by ???)
+  (Damien Doligez, report by Arseniy Alekseyev, review by Sadiq Jaffer)
 
 OCaml 4.14.1 (20 December 2022)
 ------------------------------

--- a/Changes
+++ b/Changes
@@ -8,6 +8,10 @@ OCaml 4.14 maintenance version
 
 ### Bug fixes:
 
+- #12264, #12289: Fix compact_allocate to avoid a pathological case
+  that causes very slow compaction.
+  (Damien Doligez, report by Arseniy Alekseyev, review by ???)
+
 OCaml 4.14.1 (20 December 2022)
 ------------------------------
 

--- a/runtime/compact.c
+++ b/runtime/compact.c
@@ -123,15 +123,13 @@ static void init_compact_allocate (void)
   compact_fl_small = compact_fl_large = caml_heap_start;
 }
 
-#define Chunk_free(c) (Chunk_size(c) - Chunk_alloc(c))
-
 /* [size] is a number of bytes and includes the header size */
 static char *compact_allocate (mlsize_t size)
 {
   char *chunk, *adr;
 
   chunk = size > COMPACT_FL_CUTOFF_SIZE ? compact_fl_large : compact_fl_small;
-  while (Chunk_free (chunk) < size){
+  while (Chunk_size(chunk) - Chunk_alloc(chunk) < size){
     chunk = Chunk_next (chunk);
     CAMLassert (chunk != NULL);
   }

--- a/runtime/compact.c
+++ b/runtime/compact.c
@@ -106,7 +106,12 @@ void caml_invert_root (value v, value *p)
   invert_pointer_at ((word *) p);
 }
 
-static char *compact_fl;
+static char *compact_fl_small, *compact_fl_large;
+#define COMPACT_FL_CUTOFF_SIZE Bhsize_wosize(10)
+/* COMPACT_FL_CUTOFF_SIZE should be "small" (at most 4096), as it is the
+   maximum size of "wasted" free space at the end of each chunk, and it
+   should be "large" (at least Bhsize_wosize(5)) so that the compactor
+   finds enough live blocks to fill those spaces. */
 
 static void init_compact_allocate (void)
 {
@@ -115,23 +120,23 @@ static void init_compact_allocate (void)
     Chunk_alloc (ch) = 0;
     ch = Chunk_next (ch);
   }
-  compact_fl = caml_heap_start;
+  compact_fl_small = compact_fl_large = caml_heap_start;
 }
+
+#define Chunk_free(c) (Chunk_size(c) - Chunk_alloc(c))
 
 /* [size] is a number of bytes and includes the header size */
 static char *compact_allocate (mlsize_t size)
 {
   char *chunk, *adr;
 
-  while (Chunk_size(compact_fl) - Chunk_alloc(compact_fl) < Bhsize_wosize(1)){
-    compact_fl = Chunk_next (compact_fl);
-    CAMLassert (compact_fl != NULL);
-  }
-  chunk = compact_fl;
-  while (Chunk_size (chunk) - Chunk_alloc (chunk) < size){
+  chunk = size > COMPACT_FL_CUTOFF_SIZE ? compact_fl_large : compact_fl_small;
+  while (Chunk_free (chunk) < size){
     chunk = Chunk_next (chunk);
     CAMLassert (chunk != NULL);
   }
+  *(size > COMPACT_FL_CUTOFF_SIZE ? &compact_fl_large : &compact_fl_small)
+    = chunk;
   adr = chunk + Chunk_alloc (chunk);
   Chunk_alloc (chunk) += size;
   return adr;

